Mingw-w64 warning fixes

pavel-t opened the pull request.

pavel-t committed:
Fix clobbering warning

pavel-t committed:
Avoid pointer overflow warning in wxPropertyGridPageState::DoRemoveFromSelection()

Looks like gcc8.1 think 'sel' can be empty after initialization. Avoid this by copying only the remaining entries.

pavel-t committed:
Avoid unsafe casts to enum type

pavel-t committed:
Use memcpy instead of strncpy in wxStrlcpy

Avoid stringop-overflow warning about strncpy size argument depending on source size.
Checking for null in src is not needed because it is already checked by strlen.

vadz commented on src/common/imagpng.cpp:
```diff
@@ -256,7 +256,9 @@ wxPNGHandler::LoadFile(wxImage *image,
-    unsigned char **lines = NULL;
+    // allocate the lines pointer dynamically to avoid clobbering by longjmp
+    unsigned char *** const lines_p = new unsigned char ** (NULL);
+    unsigned char ** &lines = *lines_p;
```
Sorry, I don't understand what's going on here and how does this actually help. What happens with `lines_p` after `longjmp()`?

pavel-t replied:
Local variable's value is indeterminate after longjmp, if it was modified after setjmp (https://en.cppreference.com/w/cpp/utility/program/setjmp). In practice this means that setjmp/longjmp can save/restore registers even if they are used to store local variables, so longjmp could reset lines to null and cause a memory leak. lines_p is never modified, so longjmp will just overwrite it with the same value.

vadz replied:
After rereading this once again, I realize that you probably rely on `lines_p` being unchanged because it is `const`. But I'm not actually sure if this really works like this... Are you?

I think it would be safer to have a helper function and pass it a pointer to `lines` instead. Because even if this code actually works, it's rather mysterious and I'd like to find a simpler way of fixing the problem.

vadz reviewed (commented):
Thanks for the changes, but I'd like to understand better the one in PNG loading code before applying it.

Alternatively, we could merge the other commits and leave this one out, if you prefer.

vadz committed:
Fix clobbering warnings in PNG image loading code in a better way

Add wxPNGImageData and use it to store variables that used to be local
in wxPNGHandler::LoadFile() and could be clobbered by the use of
longjmp() and pass these variables to DoLoadPNGFile(), which still uses
longjmp(), whereas LoadFile() doesn't any longer.

In addition to fixing the warning, this allows to use C++ dtors for
freeing memory and makes the code shorter and, arguably, more clear than
the original version (although written in an unusual way) and definitely
more clear than the version with the warning fix.

This reverts commit 99ea06203aec0ebb1f8978cb34a50483e0fa746c ("Fix
clobbering warning") and replaces it with a different solution to the
same problem.

vadz commented:
I've implemented what I believe to be a better/more clear solution to this problem, please let me know if you see any problems with it, otherwise I'll merge this soon.

TIA!

pavel-t commented:
Looks good, but I think it would be better to have all the clean-ups in one place. I'll try to implement this.

pavel-t commented on src/common/imagpng.cpp:
```diff
@@ -394,49 +435,47 @@ wxPNGHandler::LoadFile(wxImage *image,
+        if ( image->IsOk() )
         {
-            png_destroy_read_struct( &png_ptr, &info_ptr, (png_infopp) NULL );
-            free(info_ptr);
```
I can't find this line in the new version, was it removed intentionally?

vadz replied:
Yes, I think this is not actually needed because passing `info_ptr` to `png_destroy_read_struct()` takes care of freeing it too.

vadz commented:
> Looks good, but I think it would be better to have all the clean-ups in one place. I'll try to implement this.

Sorry, what do you mean? I'd like to merge this, do you mean I shouldn't do it?

pavel-t committed:
Make DoLoadPNGFile member function of wxPNGImageData

Avoid adding "data." everywhere in DoLoadPNGFile code

pavel-t committed:
Store info_ptr and png_ptr in wxPNGImageData

Avoid having to use goto in DoLoadPNGFile and have all clean-ups in wxPNGImageData destructor.

pavel-t committed:
Keep actual number of allocated lines in numLines

vadz commented:
Merged now, thanks for working on this!
